mm2src/trading_api/src/one_inch_api/portfolio_types.rs

```diff
@@ -0,0 +1,89 @@
+//! Structs to call 1inch portfolio api
+
+#![allow(clippy::result_large_err)]
+
+use super::client::QueryParams;
+use super::errors::ApiClientError;
+use common::{def_with_opt_param, push_if_some};
+use mm2_err_handle::mm_error::MmResult;
+use mm2_number::BigDecimal;
+use serde::Deserialize;
+
+#[derive(Default)]
+pub enum DataGranularity {
+    Month,
+    Week,
+    Day,
+    FourHour,
+    Hour,
+    FifteenMin,
+    #[default]
+    FiveMin,
+}
+
+impl ToString for DataGranularity {
+    fn to_string(&self) -> String {
+        match self {
+            DataGranularity::Month => "month".to_owned(),
+            DataGranularity::Week => "week".to_owned(),
+            DataGranularity::Day => "day".to_owned(),
+            DataGranularity::FourHour => "4hour".to_owned(),
+            DataGranularity::Hour => "hour".to_owned(),
+            DataGranularity::FifteenMin => "15min".to_owned(),
+            DataGranularity::FiveMin => "5min".to_owned(),
+        }
+    }
+}
+
+/// API params builder to get OHLC price history for token pair
+/// See 1inch docs: https://portal.1inch.dev/documentation/apis/portfolio/swagger?method=get&path=%2Fintegrations%2Fprices%2Fv1%2Ftime_range%2Fcross_prices
+#[derive(Default)]
+pub struct CrossPriceParams {
+    chain_id: u64,
+    /// Base token address
+    token0_address: String,
+    /// Quote token address
+    token1_address: String,
+    /// Returned time series intervals
+    granularity: Option<DataGranularity>,
+    /// max number of time series
+    limit: Option<u32>,
+}
```

Comment:
Why would we need OHLC data, why not use [spot price API](https://portal.1inch.dev/documentation/apis/spot-price/introduction)?

Reply:
I think, OHLC prices come from real on-chain pools so probably are more relevant to our task of estimation of the source token amount (than spot prices, which apparently represent some average market value).

Reply:
> I think, OHLC prices come from real on-chain pools so probably are more relevant to our task of estimation of the source token amount (than spot prices, which apparently represent some average market value).

But in this PR, you use the average from the last 5 minutes if I understand correctly. Why not use close price of the last candlestick, this should be the latest price.

> The spot price returns a token price in fiat currency.

You can get it for the 2 tokens and calculate the relative price by division I think. But if OHLC prices is the more accurate recent data then we should use it for sure.